hw10/stonks.py

```diff
@@ -0,0 +1,165 @@
+        pattern = re.compile(r"high52weeksDisplay: '(.*?)',")
+        var = soup.find("script", string=pattern)
+        if var:
+            match = pattern.search(str(var))
+            if match:
+                high52weeks = float(match.group(1).replace(",", ""))
+        company.max_income = (high52weeks - low52weeks) / low52weeks
+
+
+async def get_dollar_course_value(markup: str) -> float:
+    soup = BeautifulSoup(markup, "lxml")
+    return float(soup.find("valute", id="R01235").value.string.replace(",", "."))
+
+
+async def parse_information_from_one_page(
+    session,
+    base_url: str,
+    addition: str,
+    company_list: list,
+    task_list: list,
+    dollar_course: float,
+    page_param: dict,
+) -> Optional[int]:
+    async with session.get(base_url + addition, params=page_param) as resp:
+        if resp.status != 200:
+            raise ConnectionError
+
+        soup = BeautifulSoup(await resp.text(), "lxml")
+        # we have 11th page which is not listed in the finando_paging class
+        # so we'll traverse through pages until we won't find the table container index-list-container
+        if not soup.find(id="index-list-container"):
+            return -1
+        table = soup.find("table", class_="table table-small")
+        table_header = table.findChild()
+
+        for raw in table_header.find_next_siblings():
+            company = CompanyInfo()
+            company_list.append(company)
+            retrieve_main_page_company_info(company, raw)
+            task_list.append(
+                asyncio.create_task(
+                    parse_company_page_info(company, session, base_url, dollar_course)
+                )
+            )
+
+
+async def write_result_to_json(company_list: list) -> None:
+    field_names = [field.name for field in fields(CompanyInfo)]
+    df = pd.DataFrame(company_list, columns=field_names)
+    df.sort_values(by=["price"], ascending=False)[0:10].to_json(
+        "company_by_price.json", orient="records"
+    )
+    df.sort_values(by=["year_growth"], ascending=False)[0:10].to_json(
+        "company_by_year_growth.json", orient="records"
+    )
+    df.sort_values(by=["max_income"], ascending=False)[0:10].to_json(
+        "company_by_max_income.json", orient="records"
+    )
+    df.sort_values(by=["P_E"])[0:10].to_json(
+        "company_by_min_p_e.json", orient="records"
+    )
+
+
+async def main_task(base_url: str, addition: str) -> None:
+    async with aiohttp.ClientSession() as session:
+        async with session.get("http://www.cbr.ru/scripts/XML_daily.asp") as resp:
+            if resp.status != 200:
+                raise ConnectionError
+            dollar_course = await get_dollar_course_value(
+                await resp.text(),
+            )
+        # we start from first page
+        page_param = {"p": 1}
+        task_list = []
+        company_list = []
+        while True:
+            res = await parse_information_from_one_page(
```

Comment:
here when we call this function we do not expect that `parse...` function would mutate company_list and task_list. it would be better to rename it or make it return new lists (without mutating outer ones), and extend lists on 140-141 lines with the results